### PR TITLE
Dynamic Dashboard: Optimize API calls

### LINKED ISSUE #12770: Optimize API calls to only load data for enabled cards
When we reload the dashboard we should skip reloading the cards that are not enabled.

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Closes: #12770 

## Description

Stops sending unnecessary network calls for dashboard cards that are not enabled. 

Changes
- For store performance, top performance and coupon cards - Stop sending reload calls when the view model is initialized. 
- DashboardViewModel
    - Add a separate method to sync essential data upon view load.
    - Reload cards only if the card is enabled.
    - Reload only newly enabled card. 
## Testing instructions

- Fresh install the app.
- Login into the app.
- If you have orders in your store, you should see the store performance and top performers cards.
- If your store is eligible for Blaze, the Blaze card should be displayed in the dashboard. 
- If you have Proxyman or any other setup to interrupt network calls, test that network requests are sent only for enabled cards. 
- Tap "Edit" and enable any other card. 
- Save and ensure that the card is added to the dashboard and the content is loaded.
- Repeat the above steps for all the cards and validate that each card is loaded as expected.
- Close the app. 
- Relaunch and ensure that the previous configuration of cards is loaded in the dashboard.
- If you have Proxyman or any other setup to interrupt network calls, test that network requests are sent only for enabled cards. 

## Screenshots

| Dashboard cards | Customize screen |
|--------|--------|
| Simulator Screenshot - iPhone 15 Pro Max - 2024-05-18 at 18 57 22 | Simulator Screenshot - iPhone 15 Pro Max - 2024-05-18 at 18 57 24 | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 1]
```diff
@@ -136,11 +141,23 @@ final class DashboardViewModel: ObservableObject {
     func reloadAllData() async {
         await withTaskGroup(of: Void.self) { group in
             group.addTask { [weak self] in
-                guard let self else { return }
-                await self.syncAnnouncements(for: self.siteID)
+                await self?.syncDashboardEssentialData()
+            }
+            group.addTask { [weak self] in
+                guard let self = self else { return }
+                await self.reloadCards(self.showOnDashboardCards)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: Super nit:

```suggestion
                guard let self else { return }
                await reloadCards(showOnDashboardCards)
```

**Review by itsmeichigo — APPROVED** [review]
LGTM :shipit:  I see that the Blaze card has to be reloaded no matter what, but that's probably because the logic for the eligibility check is part of the reloading method and it's not simple to separate it.

I'm not 100% confident that my testing was thorough and there is no potential edge case left. It'd be nice to have unit tests to guarantee the logic - please feel free to add a separate subtask for that.

**Comment by selanthiraiyan** [comment]
Thank you for the review and testing, @itsmeichigo! 🙇 

> I see that the Blaze card has to be reloaded no matter what, but that's probably because the logic for the eligibility check is part of the reloading method and it's not simple to separate it.

Good point. Yes, I think reloading Blaze card is a compromise to avoid any edge-case issues. 

> I'm not 100% confident that my testing was thorough and there is no potential edge case left. It'd be nice to have unit tests to guarantee the logic - please feel free to add a separate subtask for that.

I agree about the potential edge cases. Logged a subtask https://github.com/woocommerce/woocommerce-ios/issues/12782

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: Done in https://github.com/woocommerce/woocommerce-ios/pull/12775/commits/df5b2fc329d152f9a0416342789fa343f05f7fc8

**Pull request merged**